src/kubernetes.cc

```diff
@@ -115,6 +184,52 @@ KubernetesReader::KubernetesReader(const Configuration& config,
                                    HealthChecker* health_checker)
     : config_(config), environment_(config), health_checker_(health_checker) {}
 
+MetadataStore::Metadata KubernetesReader::GetMetadata(
+    const json::Object* object, Timestamp collected_at, bool is_deleted) const
+    throw(json::Exception) {
+#ifndef ENABLE_KUBERNETES_METADATA
+  return MetadataStore::Metadata::IGNORED();
+#else
+  const std::string cluster_location = environment_.KubernetesClusterLocation();
+
+  const std::string kind = object->Get<json::String>("kind");
+  const std::string api_version = object->Get<json::String>("apiVersion");
+  const json::Object* metadata = object->Get<json::Object>("metadata");
+  const std::string self_link = metadata->Get<json::String>("selfLink");
+  const std::pair<std::string, std::string> type_and_version =
+      TypeAndVersion(api_version, kind);
+  const std::string type = type_and_version.first;
+  const std::string version = type_and_version.second;
+
+  const std::string schema =
+      format::Substitute(kKubernetesSchemaNameFormat,
+                         {{"type", type}, {"version", version}});
+  const std::string created_str =
+      metadata->Get<json::String>("creationTimestamp");
+  Timestamp created_at = time::rfc3339::FromString(created_str);
+  const std::string resource_full_name = FullResourceName(self_link);
+
+  if (config_.VerboseLogging()) {
+    LOG(INFO) << "Raw resource metadata for full name: " << resource_full_name
+              << ": " << *object;
+  }
+  return MetadataStore::Metadata(resource_full_name, type, cluster_location,
+                                 version, schema, is_deleted, collected_at,
+                                 object->Clone());
+#endif
+}
+
+MetadataUpdater::ResourceMetadata KubernetesReader::GetResourceMetadata(
+    const json::Object* object, Timestamp collected_at, bool is_deleted) const
+    throw(json::Exception) {
+  MetadataStore::Metadata metadata =
+      GetMetadata(object, collected_at, is_deleted);
+  return MetadataUpdater::ResourceMetadata(
+      /*ids=*/std::vector<std::string>{},
+      MonitoredResource("", {}),
+      metadata.Clone());
+}
+
 MetadataUpdater::ResourceMetadata KubernetesReader::GetNodeMetadata(
```

Comment:
What I was hoping for in the refactoring is that would end up with one `GetObjectMetadata` that could be customized for nodes and pods via a callback.

Reply:
I see. Will handle in a follow up commit.